File: internal/validate/code.go
package validate

import (
	"github.com/PurnaOS/iBuildOS/internal/config"
	"github.com/PurnaOS/iBuildOS/internal/model"
	"github.com/PurnaOS/iBuildOS/internal/okf"
	"github.com/PurnaOS/iBuildOS/internal/types"
	"gopkg.in/yaml.v3"
)

// validateCode enforces the "Code" node of the chain: any artifact carrying the
// configured code field (default `code`) must have at least one glob match on
// disk. Globs are repo-relative (resolved against the bundle dir). Absence of
// the field is fine here — a done task that declares no code is handled by the
// completeness rules instead.
func validateCode(a *artifact, _ *types.Registry, cfg config.Config, c *model.Collector) {
	globs := scalarListField(a, cfg.Chain.CodeField)
	if len(globs) == 0 {
		return
	}
	kn, _, _ := a.doc.Get(cfg.Chain.CodeField)
	line := a.doc.Line(kn)
	matched, err := okf.AnyMatch(cfg.BundleDir, globs)
	if err != nil {
		c.Errf(a.path, line, "code.noMatch", "invalid code glob in %v: %v", globs, err)
		return
	}
	if !matched {
		c.Errf(a.path, line, "code.noMatch", "code globs %v matched no files on disk", globs)
	}
}

// scalarListField returns the string items of a list-valued frontmatter field.
// A scalar value is tolerated as a single-element list (one glob), so a code
// field written as `code: pkg/**` is treated the same as `code: [pkg/**]` by
// both the completeness rules and validateCode. Returns nil when the field is
// absent, empty, or a non-scalar/non-list value.
func scalarListField(a *artifact, field string) []string {
	if a.doc == nil {
		return nil
	}
	_, vn, ok := a.doc.Get(field)
	if !ok {
		return nil
	}
	switch vn.Kind {
	case yaml.ScalarNode:
		if vn.Value != "" {
			return []string{vn.Value}
		}
		return nil
	case yaml.SequenceNode:
		var out []string
		for _, item := range vn.Content {
			if item.Value != "" {
				out = append(out, item.Value)
			}
		}
		return out
	default:
		return nil
	}
}

File: src/core/validate/code.ts
// Enforces the "Code" node of the chain: an artifact carrying the configured
// code field must have at least one glob match on disk. Port of Go
// internal/validate/code.go.
import type { Config } from "../config/config.ts";
import { Collector } from "../model/model.ts";
import { anyMatch } from "../okf/glob.ts";
import { scalarText, nodeIsScalar, nodeIsSeq, seqItems } from "../okf/frontmatter.ts";
import type { Artifact } from "./validate.ts";

export function validateCode(a: Artifact, cfg: Config, c: Collector): void {
  const globs = scalarListField(a, cfg.chain.codeField);
  if (globs.length === 0) return;
  const g = a.doc!.get(cfg.chain.codeField);
  const line = a.doc!.line(g?.keyNode);
  let matched: boolean;
  try {
    matched = anyMatch(cfg.bundleDir, globs);
  } catch (e) {
    c.errf(a.path, line, "code.noMatch", `invalid code glob in ${JSON.stringify(globs)}: ${(e as Error).message}`);
    return;
  }
  if (!matched) {
    c.errf(a.path, line, "code.noMatch", `code globs ${JSON.stringify(globs)} matched no files on disk`);
  }
}

// scalarListField returns the string items of a list-valued frontmatter field. A
// scalar value is tolerated as a single-element list. nil when absent/empty.
export function scalarListField(a: Artifact, field: string): string[] {
  if (!a.doc) return [];
  const g = a.doc.get(field);
  if (!g) return [];
  const vn = g.valNode;
  if (nodeIsScalar(vn)) {
    const t = scalarText(vn);
    return t !== "" ? [t] : [];
  }
  if (nodeIsSeq(vn)) {
    return seqItems(vn).map(scalarText).filter((t) => t !== "");
  }
  return [];
}
